amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp
+
+import java.util.concurrent.atomic.AtomicReference
+
+import com.rabbitmq.client.Connection
+
+import scala.compat.java8.FunctionConverters._
+
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionProvider {
+  def settings: AmqpConnectionSettings
+  def get: Connection
+  def release(connection: Connection): Unit
+}
+
+final object DefaultAmqpConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings): DefaultAmqpConnectionProvider =
+    DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings)
+}
+
+final case class DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings) extends AmqpConnectionProvider {
+  override def get = settings.getConnection
```

Comment:
Hmmm, shouldn't the connection logic go here rather than live in the settings now that we have a `provider`?

Reply:
It could but then each provider will need to match against all possible settings.
It's more reliable if each settings can provide its own connection so all providers work with all settings.

Other option is what I did in my original PR. Get rid of settings and consider them as providers and the reusable provider just a provider that wrap other providers.

Reply:
Hmm, maybe it's more of a naming thing, but, now we have a provider that does not provide anything but has a special case where it caches connections, and a `Settings` that creates/provides connections, this makes me think we are missing something with the modeling ;)